### PR TITLE
Hw4

### PR DESCRIPTION
**Pull request by YuriKitt**
**Task:**
Test Setup:

- Set up a testing environment using Jest to write and run unit tests.
- Ensure that your project's dependencies include Jest and related testing libraries.

Unit Tests for APIs:

- Write unit tests for the API endpoints related to "Movies" and "Genres."
- Cover scenarios such as creating, reading, updating, and deleting movies and genres.
- Include tests for error handling and validation of input data.

Unit Tests for Movie Genre Search:

- Write unit tests for the movie genre search functionality.
- Test different genres and ensure that the API returns the expected results.
- Include edge cases and error scenarios in your tests.

Coverage and Reporting:

- Aim for high code coverage in your tests to ensure thorough testing of your application.
- Generate and include a code coverage report as part of your test results.

Documentation:

- Update your project's documentation to include information on running the tests using Jest.
- Provide clear instructions on how to run the tests and interpret the test results.

_Please review._

### CODE REVIEW

**Review by IhorOsadchyi — CHANGES_REQUESTED** [review]
As for me, using database for unit tests is not a right way. It's more like integration tests or even end-to-end. For unit testing you need to isolate function or method from 3-d party APIs using mock data. If you want, try to mock DB responses for testing purposes.

**Comment by YuriKitt** [comment]
> As for me, using database for unit tests is not a right way. It's more like integration tests or even end-to-end. For unit testing you need to isolate function or method from 3-d party APIs using mock data. If you want, try to mock DB responses for testing purposes.

all responses have been replaced with mock data.

File: src/tests/genre.test.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+describe('GET /api/genres/:id', () => {
+  it('should return 404 if genre is not found', async () => {
+    const nonExistentId = 'aaaa12345678901234567890';
+    Genre.findById = jest.fn().mockResolvedValue(null);
+    const res = await request(app).get(`/api/genres/${nonExistentId}`);
+    expect(Genre.findById).toHaveBeenCalledWith(nonExistentId);
+    expect(res.body.error).toEqual('Genre not found');
+    expect(res.statusCode).toEqual(404);
+  });
+});
```
Comment (IhorOsadchyi) [inline src/tests/genre.test.ts]: There test cases are related to same endpoint so you can put them inside same describe block.

File: src/tests/movie.test.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+      .put(`/api/movies/${nonExistentId}`)
+      .send(updatedData);
+    expect(Movie.findByIdAndUpdate).toHaveBeenCalledWith(
+      nonExistentId,
+      updatedData,
+      { new: true }
+    );
+    expect(res.statusCode).toEqual(404);
+  });
+});
```
Comment (IhorOsadchyi) [inline src/tests/movie.test.ts]: The same situation as above.

File: src/tests/genre.test.ts (hunk 1 above)
Comment (YuriKitt) [reply src/tests/genre.test.ts]: resolved

File: src/tests/movie.test.ts (hunk 2 above)
Comment (YuriKitt) [reply src/tests/movie.test.ts]: resolved

**Review by IhorOsadchyi — APPROVED** [review]
